source/Irrlicht/CIrrDeviceLinux.cpp

```diff
@@ -993,49 +995,104 @@ bool CIrrDeviceLinux::run()
+						response.xselection.property = property;
+						response.xselection.time = req->time;
+						XSendEvent (XDisplay, req->requestor, 0, 0, &response);
+						XFlush (XDisplay);
+					};
+					auto send_response_refuse = [&send_response] {
+						send_response(None);
+					};
+
+					// sets the required property to data of type type and
+					// sends the according response
+					auto set_property_and_notify = [this, req, &send_response]
+							(Atom type, int format, const void *data, u32 data_size) {
+						XChangeProperty(XDisplay,
 								req->requestor,
-								req->property, req->target,
-								8, // format
+								req->property,
+								type,
+								format,
 								PropModeReplace,
-								(unsigned char*) Clipboard.c_str(),
-								Clipboard.size());
-						respond.xselection.property = req->property;
+								(const unsigned char *)data,
+								data_size);
+						send_response(req->property);
+					};
+
+					if (req->selection != X_ATOM_CLIPBOARD ||
+							req->owner != XWindow) {
+						// we are not the owner, refuse request
+						send_response_refuse();
+						break;
 					}
-					else if ( req->target == X_ATOM_TARGETS )
-					{
-						long data[2];
 
-						data[0] = X_ATOM_TEXT;
-						data[1] = XA_STRING;
-
-						XChangeProperty (XDisplay, req->requestor,
-								req->property, req->target,
-								8, PropModeReplace,
-								(unsigned char *) &data,
-								sizeof (data));
-						respond.xselection.property = req->property;
+					// for debugging:
+					//~ {
+						//~ char *target_name = XGetAtomName(XDisplay, req->target);
+						//~ fprintf(stderr, "CIrrDeviceLinux::run: target: %s (=%ld)\n",
+								//~ target_name, req->target);
+						//~ XFree(target_name);
+					//~ }
+
+					if (req->property == None) {
+						// req is from obsolete client, use target as property name
+						// and X_ATOM_UTF8_STRING as type
+						// Note: this was not tested and might be incorrect
+						os::Printer::log("CIrrDeviceLinux::run: SelectionRequest from obsolete client",
+								ELL_WARNING);
+						XChangeProperty(XDisplay,
+								req->requestor,
+								req->target, X_ATOM_UTF8_STRING,
+								8, // format = 8-bit
+								PropModeReplace,
+								(unsigned char *)Clipboard.c_str(),
+								Clipboard.size());
+						send_response(req->target);
+						break;
 					}
-					else
-					{
-						respond.xselection.property= None;
+
+					if (req->target == X_ATOM_TARGETS) {
+						Atom data[] = {
+							X_ATOM_TARGETS,
+							X_ATOM_TEXT,
+							X_ATOM_UTF8_STRING,
+							X_ATOM_UTF8_MIME_TYPE
+						};
+						set_property_and_notify(
+								XA_ATOM,
+								32, // Atom is long, we need to set 32 for longs
+								&data,
+								sizeof(data) / sizeof(*data)
+							);
+
+					} else if (req->target == X_ATOM_TEXT ||
+							req->target == X_ATOM_UTF8_STRING ||
+							req->target == X_ATOM_UTF8_MIME_TYPE) {
+						set_property_and_notify(
+								X_ATOM_UTF8_STRING,

```

Comment:
If I understand it correctly, target != type, so `UTF8_STRING` should probably be a correct type for `text/plain;charset=utf-8`.
There's a mapping for some targets to its types: https://www.x.org/releases/current/doc/xorg-docs/icccm/icccm.html#Target_Atoms But sadly it doesn't include any utf-8 stuff.

The TEXT target should be correct:
https://www.x.org/releases/current/doc/xorg-docs/icccm/icccm.html#TEXT_Properties
>The atom TEXT is a polymorphic target. Requesting conversion into TEXT will convert into whatever encoding is convenient for the owner. The encoding chosen will be indicated by the type of the property returned. TEXT is not defined as a type; it will never be the returned type from a selection conversion request.

`XA_STRING` is "ISO Latin-1 (+TAB+NEWLINE) text"(https://www.x.org/releases/current/doc/xorg-docs/icccm/icccm.html#Target_Atoms) I don't think trying to convert our utf-8 strings to ISO Latin-1 would give us any benefit.
I've seen at least one application (wezterm) that just uses `UTF8_STRING` (wezterm), so we wouldn't be the only one to not support those very old applications, of which I haven't seen any while testing, btw..

There is a benefit from using `X_ATOM_UTF8_MIME_TYPE` though: There are some apps (ie. xpad) that seem to ignore what the `TARGETS` target says and just query some targets one after another. Because minetest runs the event handling stuff with a lower frequency when unfocused (which is also when these requests happen), you can see a delay for those apps. And because xpad requests `text/plain;charset=utf-8`(=`X_ATOM_UTF8_MIME_TYPE`), and then `UTF8_STRING` (no matter what `TARGETS` says), I've added support for`X_ATOM_UTF8_MIME_TYPE` to slightly reduce the delay.

Reply:
Sounds okay.